java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOutputListener.java

```diff
@@ -45,67 +43,125 @@ public BazelJUnitOutputListener(Path xmlOut) {
+      TestData parent = testCase.getId().getParentIdObject().map(results::get).orElse(null);
+      if (parent == null) {
+        // We should really log this, because something has gone wildly wrong
+        continue;
+      }
+
+      knownSuites.computeIfAbsent(parent, id -> new ArrayList<>()).add(testCase);
+    }
+
+    return knownSuites;
+  }
+
+  private List<TestData> findTestCases() {
+    return results.values().stream()
+        // Ignore test plan roots. These are always the engine being used.
+        .filter(result -> !testPlan.getRoots().contains(result.getId()))
+        .filter(
+            result -> {
+              // Find the test results we will convert to `testcase` entries. These
+              // are identified by the fact that they have no child test cases in the
+              // test plan, or they are marked as tests.
+              TestIdentifier id = result.getId();
+              return id.isTest() || testPlan.getChildren(id).isEmpty();
+            })
+        .collect(Collectors.toList());
   }
 
   @Override
   public void dynamicTestRegistered(TestIdentifier testIdentifier) {
-    roots.forEach(root -> root.addDynamicTest(testIdentifier));
+    getResult(testIdentifier).setDynamic(true);
   }
 
   @Override
-  public void executionStarted(TestIdentifier testIdentifier) {
-    roots.forEach(root -> root.markStarted(testIdentifier));
+  public void executionSkipped(TestIdentifier testIdentifier, String reason) {
+    getResult(testIdentifier).mark(TestExecutionResult.aborted(null)).skipReason(reason);
+    outputIfTestRootIsComplete(testIdentifier);
   }
 
   @Override
-  public void executionSkipped(TestIdentifier testIdentifier, String reason) {
-    roots.forEach(root -> root.markSkipped(testIdentifier, reason));
+  public void executionStarted(TestIdentifier testIdentifier) {
+    getResult(testIdentifier).started();
   }
 
   @Override
   public void executionFinished(
       TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
-    roots.forEach(root -> root.markFinished(testIdentifier, testExecutionResult));
+    getResult(testIdentifier).mark(testExecutionResult);
+    outputIfTestRootIsComplete(testIdentifier);
+  }
+
+  private void outputIfTestRootIsComplete(TestIdentifier testIdentifier) {
+    if (!testPlan.getRoots().contains(testIdentifier)) {
+      return;
+    }
+
+    List<TestData> testCases = findTestCases();
+    Map<TestData, List<TestData>> testSuites = matchTestCasesToSuites(testCases);
+
+    // Write the results
+    try {
+      for (Map.Entry<TestData, List<TestData>> suiteAndTests : testSuites.entrySet()) {
+        new TestSuiteXmlRenderer(testPlan)
+            .toXml(xml, suiteAndTests.getKey(), suiteAndTests.getValue());
+      }
+    } catch (XMLStreamException e) {
+      throw new RuntimeException(e);
+    }
+
+    // Delete the results we've used to conserve memory
+    Stream.concat(testCases.stream(), testSuites.keySet().stream())
```

Comment:
What guarantees do we have that the suites are complete at this point and won't have more child test-cases which reference them? Given the code in `matchTestCasesToSuites` ignores any tests where we don't know about their containing suite, I'm quite wary of throwing away test suites unless we're very sure we're done with them (which could do with a comment justifying why we we're so sure :))